### PR TITLE
add expiry time to registration entry and pruning API

### PR DESCRIPTION
**Pull request by amoore877**
Signed-off-by: Andrew Moore <andrew.s.moore@uber.com>

**Pull Request check list**

- [X] Commit conforms to CONTRIBUTING.md?
- [X] Proper tests/regressions included?
- [X] Documentation updated?

**Affected functionality**
`PruneJoinTokens` changed to `<` not `<=` to match naming of `ExpiresBefore` field and the existing documentation of the functionality.
New functionality: Expiry field to `RegistrationEntry` and `Prune` API for the same.

**Description of change**
Add **optional** Expiry field to `RegistrationEntry`, analogous to the same in `JoinToken`. Add `Prune` API for `RegistrationEntry`, also analogous to the same in `JoinToken`. Add related tests for updating model and for the new API.

**Which issue this PR fixes**
Addresses base need of #773. Discussion still on-going about additional logic/features to add utilizing the new API/information. Does **not** include a scheduled routine for pruning.

### CODE REVIEW

File: pkg/server/plugin/datastore/sql/sql.go [hunk 1]
```diff
@@ -1079,19 +1093,43 @@ func deleteRegistrationEntry(tx *gorm.DB,
...
+	if err := tx.Delete(&entry).Error; err != nil {
+		return sqlError.Wrap(err)
+	}
+
+	return nil
+}
+
+func pruneRegistrationEntries(tx *gorm.DB, req *datastore.PruneRegistrationEntriesRequest) (*datastore.PruneRegistrationEntriesResponse, error) {
+	var registrationEntries []RegisteredEntry
+	if err := tx.Where("expiry != 0").Where("expiry <= ?", req.ExpiresBefore).Find(&registrationEntries).Error; err != nil {
```
Comment (azdagron) [inline pkg/server/plugin/datastore/sql/sql.go]: the field name ExpiresBefore suggests this should be a `<`?

File: test/fakes/fakedatastore/fakedatastore.go [hunk 2]
```diff
@@ -494,6 +494,19 @@ func (s *DataStore) DeleteRegistrationEntry(ctx context.Context,
 	}, nil
 }
 
+func (s *DataStore) PruneRegistrationEntries(ctx context.Context, req *datastore.PruneRegistrationEntriesRequest) (*datastore.PruneRegistrationEntriesResponse, error) {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+
+	for key, entry := range s.registrationEntries {
+		if entry.Expiry <= req.ExpiresBefore {
```
Comment (azdagron) [inline test/fakes/fakedatastore/fakedatastore.go]: This does not match the real prune semantics (does not check for `== 0`)

File: pkg/server/plugin/datastore/sql/sql.go (hunk 1 above)
Comment (amoore877) [reply pkg/server/plugin/datastore/sql/sql.go]: I was wondering about that; `pruneJoinTokens` has this same logic, so I think both should be changed to `<` not `<=`. Would that be alright?

Comment (amoore877) [reply pkg/server/plugin/datastore/sql/sql.go]: pushed change assuming "yes" to above :)

**Pull request merged**